### ISSUE TITLE
Localstack Lambda won't return Tail log on invoke

### ISSUE THREAD

**Issue by roman-io** [localstack/localstack#1630]
MacOS 10.14.6 Mojave.
Node v10.15.3
LocalStack - docker image 0.10.4

When invoking a Lambda from Node.js code.
```
const lambda = new Aws.Lambda(creds);

const payload = {};

const clientContext = {
    invokedFunctionArn: this._workerArn
};

const params = {
    FunctionName: this._workerArn,
    InvocationType: "RequestResponse", // "DryRun",
    LogType: 'Tail', // "None", "Tail"
    ClientContext: JSON.stringify(clientContext),
    Payload: JSON.stringify(payload)
};

console.log(`DEBUG: Executing lambda ${JSON.stringify(params)}`);

lambda.invoke(params, function (err, data) {
    console.log(`DEBUG: Printing lambda return arguments`);
    console.log(JSON.stringify(arguments));
    console.log(data.LogResult);
});
```

```
DEBUG: Printing lambda return arguments
{"0":null,"1":{"StatusCode":200,"Payload":"null"}}
undefined
```

Expecting to see the Lambda execution log in `data.LogResult`, but no such object returned,
yet Localstack Cloudwatch logs recorded the execution logs correctly.
```
EVENTS	75	1570381308926.664	+51733/05/27/[$LATEST]629146bb	START RequestId: dbd86068-78c8-1a62-29ca-9c0a4a8580c3 Version: $LATEST	1570381306039
EVENTS	76	1570381308926.664	+51733/05/27/[$LATEST]629146bb	2019-10-06T17:01:47.924Z	dbd86068-78c8-1a62-29ca-9c0a4a8580c3	INFO	DEBUG: LOCALSTACK_HOSTNAME=172.17.0.1	1570381306520
...
EVENTS	79	1570381308926.664	+51733/05/27/[$LATEST]629146bb	END RequestId: dbd86068-78c8-1a62-29ca-9c0a4a8580c3	1570381307961
EVENTS	80	1570381308926.664	+51733/05/27/[$LATEST]629146bb	REPORT RequestId: dbd86068-78c8-1a62-29ca-9c0a4a8580c3	Duration: 16.05 ms	Billed Duration: 100 ms	Memory Size: 1536 MB	Max Memory Used: 57 MB		1570381308442
```

What I'm missing? Why it doesn't return logs on invoke?

Thanks

**Comment by jarrodallan** [comment]
I am facing the same issue when attempting similar operations using the AWS CLI rather than python to deploy and invoke the lambda.

I have followed the AWS guide [here](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-awscli.html) which uses the AWS CLI to create IAM policies and attach these to a new role, then use this role in the creation of a lambda. Each of the create operations are successful.
When invoking the lambda using the CLI the `LogResult` is not returned when using the `LogType: Tail`.

Using the CLI I can see that the CloudWatch Log Group and a Log Stream are created when invoked, however the tail is not working when used with the `invoke` command.

I have set the environment variable `DEFAULT_REGION=us-east-1`and `AWS_DEFAULT_REGION=us-east-1`, as well as setting it in the CLI command with `--region=us-east-1` and in the configured profile as well.

When following the guide exactly using a real AWS account the `Tail` is successful as specified in the guide.

```
aws --version
aws-cli/1.18.23 Python/3.8.2 Darwin/19.4.0 botocore/1.15.4

localstack --version
0.11.2
```

**Comment by cristian-marinescu** [comment]
Same here. The IAM role attached has the appropriate policies. Logs are recorded to CloudWatch, but direct invocation with `aws --endpoint http://localhost:4566 lambda invoke --function-name <f_name> out.txt --log-type Tail` does not produce logs, `LogResult` is an empty string.
`
{
    "StatusCode": 200,
    "LogResult": "",
    "ExecutedVersion": "$LATEST"
}
`

`
localstack_main | LocalStack version: 0.11.3
aws --version
aws-cli/2.0.17 Python/3.8.3 Darwin/19.5.0 botocore/2.0.0dev21
`
My docker-compose.yml:
`
version: '3'
services:
  localstack:
    container_name: "${LOCALSTACK_DOCKER_NAME-localstack_main}"
    image: localstack/localstack
    environment:
      - SERVICES=sts,iam,lambda,dynamodb,s3,stepfunctions,cloudformation,cloudwatch
    ports:
      - "4566-4599:4566-4599"
      - "${PORT_WEB_UI-8080}:${PORT_WEB_UI-8080}"
    volumes:
      - "${TMPDIR:-/tmp/localstack}:/tmp/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
`
The exact same configuration works fine when deploying to AWS.

**Comment by jarrodallan** [comment]
@cristian-marinescu - What region are you trying to deploy to when you use localstack? I have had some issue retrieving logs even using `aws logs` unless I explicitly set `region: us-east-1`

**Comment by cristian-marinescu** [comment]
@jarrodallan I should have mentioned this, `us-east-1` was used implicitly. It is the default region. The behaviour I described is agnostic of explicit/implicit passing of region in any of the commands.

**Comment by jarrodallan** [comment]
That was the same behaviour I have experienced just wanted to confirm.

**Comment by BenRbtz** [comment]
This doesn't seem to be implemented currently. I've come to this conclusion from these points:
1. A TODO to add LogResult in the header:
https://github.com/localstack/localstack/blob/master/localstack/services/awslambda/lambda_api.py#L1364
2. Couldn't find any reference of it being set while looking through the source (`grep 'LogResult' -ir .`)
3. Tests don't use LogResult to check the lambda output but instead using log-groups: https://github.com/localstack/localstack/blob/5a723c368b317f3dc61952709c1d07af56be332e/tests/integration/test_lambda.py#L828
4. testing on the latest localstack image

(Please don't take this conclusion as gospel. This was just my findings while looking through the source)

I have three suggestions as workarounds:
1. If your lambdas are hosted in a docker container: `docker container logs $containerName`
2. Use aws-cli to invoke the lambda. The output file produced will have the lambda output: `aws --endpoint-url=http://localhost:4566 lambda invoke --function-name lambda_foo output_file.txt `
3. Using CloudWatch log groups like shown in the test_lambda.py

**Comment by BenRbtz** [comment]
@whummer I can see you've merged a fix already - thanks for picking this up!

I've pulled the new latest image and it now returns a LogResult as expected. However, in the case of the lambda raising an error, it will return LogResult with an empty value..

**Comment by BenRbtz** [comment]
After further testing, it seems that it doesn't capture output from logging module.

**Comment by BenRbtz** [comment]
To ensure that I have written the Lambda correctly and understand how AWS Lambda handles log capturing, I created the following Python Lambda in both AWS and in Localstack:

```python
import logging

logging.getLogger().setLevel(logging.INFO)


def lambda_handler(event, context):
    print('FOO')
    logging.info('BAR')
    return 'OK'
```

AWS Lambda output:
```
START RequestId: 5077bac3-9a76-4a2e-99fb-ca115ce3acc8 Version: $LATEST
FOO
[INFO]	2020-12-31T08:54:32.217Z	5077bac3-9a76-4a2e-99fb-ca115ce3acc8	BAR
END RequestId: 5077bac3-9a76-4a2e-99fb-ca115ce3acc8
REPORT RequestId: 5077bac3-9a76-4a2e-99fb-ca115ce3acc8	Duration: 1.22 ms	Billed Duration: 2 ms	Memory Size: 128 MB	Max Memory Used: 52 MB	Init Duration: 120.83 ms
```

Localstack Lambda output:
```
START 07e5a0d4-78eb-4dd3-9c62-21b0ed96d8cb: Lambda arn:aws:lambda:eu-west-1:000000000000:function:lambda_bar started via "local" executor ...
FOO
END RequestId: 07e5a0d4-78eb-4dd3-9c62-21b0ed96d8cb
REPORT RequestId: 07e5a0d4-78eb-4dd3-9c62-21b0ed96d8cb Duration: 0 ms
```

As you can seen from the outputs, I can confirm there is something causing Localstack Lambda not to display `BAR` from the logging module. After some research, which could be unrelated, it seems that there's [some bootstrapping](https://stackoverflow.com/a/56579088) in the AWS version of the Lambda for logging. Possibly we want something similar in Localstack, if not already implemented?

I've tried to work around this by setting a new logger i.e. `logging.basicConfig(level=logging.INFO)` but to no avail. 

Also, I can confirm that the AWS Lambda does capture logs for a thrown error.

**Comment by whummer** [comment]
Thanks for the update @BenRbtz . Looks like you're using `LAMBDA_EXECUTOR=local` in your configuration, is that right? The problem is that running Python Lambdas with `local` are running in the same process as the LocalStack code, which can interfere with the logging configuration. Did you have a chance to try it with the default Docker-based Lambda executor?

> Also, I can confirm that the AWS Lambda does capture logs for a thrown error.

Just to confirm - the logs are now correctly returned when an error is thrown, right?

**Comment by BenRbtz** [comment]
>Looks like you're using LAMBDA_EXECUTOR=local in your configuration, is that right?

Nope, didn't set that in the configuration as I thought that was default. Checked the docs and it was failing over to use it instead. 

> Did you have a chance to try it with the default Docker-based Lambda executor?

Just fixed the `docker-compose.yml` and tried the Docker-based Lambda executor and that has fixed the missing logging issue, thanks for the suggestion!

> Just to confirm - the logs are now correctly returned when an error is thrown in the Lambda, right?

Sorry for my initial response not being very clear. The current behaviour for the lambda logs are:
* When the Lambda doesn't throw an errors - logs in response
* When the Lambda does throw an error - no logs in response
